build/curriculum.ts

```diff
@@ -0,0 +1,422 @@
+      .replace(/ modules$/, "") // Strip "modules" suffix.
+      .replace(/Extension \d+:/, ""); // Strip "Extension" prefix.
+  }
+
+  return {
+    meta: {
+      filename,
+      slug,
+      url,
+      title,
+      sidebar,
+      modules,
+      parents,
+      prevNext,
+      ...attributes,
+    },
+    body,
+  };
+}
+
+export async function findCurriculumPageBySlug(
+  slug: string
+): Promise<CurriculumData | null> {
+  let file = await slugToFile(slug);
+  if (!file) {
+    file = await slugToFile(`${slug}${slug ? "/" : ""}README`);
+  }
+  let module;
+  try {
+    module = await readCurriculumPage(file, { forIndex: false });
+  } catch (e) {
+    console.error(`No file found for ${slug}: ${e}`);
+    return;
+  }
+  const { body, meta } = module;
+
+  const d: CurriculumBuildData = {
+    url: meta.url,
+    rawBody: body,
+    metadata: { locale: DEFAULT_LOCALE, ...meta },
+    isMarkdown: true,
+    fileInfo: {
+      path: file,
+    },
+  };
+
+  const doc = await buildCurriculumPage(d);
+  return { doc };
+}
+
+export async function buildCurriculumPage(
+  document: CurriculumBuildData
+): Promise<Doc> {
+  const { metadata } = document;
+
+  const doc = { locale: DEFAULT_LOCALE } as Partial<CurriculumDoc>;
+  let $ = null;
+
+  const renderUrl = document.url.replace(/\/$/, "");
+  [$] = await kumascript.render(renderUrl, {}, document as any);
+
+  $("[data-token]").removeAttr("data-token");
+  $("[data-flaw-src]").removeAttr("data-flaw-src");
+
+  doc.title = (metadata.title || "").replace(/^\d+\s+/, "");
+  doc.mdn_url = document.url;
+  doc.locale = metadata.locale as string;
+  doc.native = LANGUAGES_RAW[DEFAULT_LOCALE]?.native;
+
+  if ($("math").length > 0) {
+    doc.hasMathML = true;
+  }
+  $("div.hidden").remove();
+  syntaxHighlight($, doc);
+  injectNoTranslate($);
+  injectLoadingLazyAttributes($);
+  postProcessCurriculumLinks($, (p: string) => {
+    const [head, hash] = p.split("#");
+    const slug = fileToSlug(
+      path.normalize(path.join(path.dirname(document.fileInfo.path), head))
+    ).replace(/\/$/, "");
+    return `/${DEFAULT_LOCALE}/${slug}/${hash ? `#${hash}` : ""}`;
+  });
+  postProcessExternalLinks($);
+  postLocalFileLinks($, doc);
+  postProcessSmallerHeadingIDs($);
+  wrapTables($);
+  setCurriculumTypes($);
+  try {
+    const [sections] = await extractSections($);
+    doc.body = sections;
+  } catch (error) {
+    console.error(
+      `Extracting sections failed in ${doc.mdn_url} (${document.fileInfo.path})`
+    );
+    throw error;
+  }
+
+  doc.pageTitle = `${doc.title} | MDN Curriculum`;
+
+  doc.noIndexing = false;
+  doc.toc = makeTOC(doc, true).map(({ text, id }) => {
+    return { text: text.replace(/^[\d.]+\s+/, ""), id };
+  });
+  doc.sidebar = metadata.sidebar;
+  doc.modules = metadata.modules;
+  doc.prevNext = metadata.prevNext;
+  doc.parents = metadata.parents;
+  doc.topic = metadata.topic;
+
+  return doc as Doc;
```

Comment:
What we return here doesn't look like a `Doc`:

https://github.com/mdn/yari/blob/3d8dbb467ce2ef1a321938312daeb1ed8d9406a8/libs/types/document.ts#L154-L159

If we have no choice right now, let's at least add a `FIXME` comment to avoid future confusion.

Reply:
Agreed, this isn't a Doc, it should be a CurriculumDoc, but we shouldn't use an assertion here, best to assemble the correct type here:
```
  return {
    title: doc.title,
    mdn_url: doc.mdn_url,
  ...
```

or never use a partial type:
```
  const doc: CurriculumDoc = {
    title: (metadata.title || "").replace(/^\d+\s+/, ""),
    mdn_url: document.url,
  ...
```

Reply:
This very close to `buildDocument` from content and they also do the cast. It's fragile and cannot just be constructed at the end.